lisp/ess-developer.el

```diff
@@ -30,130 +30,248 @@
 
 ;;; Code:
 
-;; (require 'ess-site) ;; need to assigne the keys in the map
-
 (defgroup ess-developer nil
   "ESS: developer."
   :group 'ess
   :prefix "ess-developer-")
 
-(defface ess-developer-indicator-face
-  '((((class grayscale)) (:background "DimGray"))
-    (((class color) (background light))
-     (:foreground "red4"  :bold t ))
-    (((class color) (background dark))
-     (:foreground "deep sky blue"  :bold t )))
-  "Face to highlight mode line process name when developer mode is on."
-  :group 'ess-developer)
-
-(defcustom ess-developer-packages nil
-  "List of names of R packages you currently develop.
-Set this variable to the list of packages you commonly develop or
-use `ess-developer-add-package' to modify interactively this
-list."
-  :group 'ess-developer
-  :type 'list)
+(defvar ess-developer-local-package nil
+  "Cons cell of two strings. The first contains the package name
+active in the current buffer, and the second is the path to that
+package source directory.")
+(make-variable-buffer-local 'ess-developer-local-package)
 
-(defcustom ess-developer-load-package-command "library(devtools)\nload_all('%s')\n"
-  "Command issued by `ess-developer-load-package'.
- %s is subsituted with the user supplied directory."
+(defvar ess-developer-code-injection nil
+  "String with the package name where source code should be
+injected.")
+(make-variable-buffer-local 'ess-developer-code-injection)
+
+(defcustom ess-developer-code-injection-in-packages nil
+  "If non-nil, `ess-developer-code-injection' is automatically
+set within R packages."
   :group 'ess-developer
-  :type 'string)
+  :type 'boolean)
 
 (defvar ess-developer-root-file "DESCRIPTION"
   "If this file is present in the directory, it is considered a
-  project root.")
-
-;; (defcustom ess-developer-force-attach nil
-;;   "If non-nill all the packages listed in `ess-developer-packages' should be attached
-;; when ess-developer mode is turned on."
-;;   :group 'ess-developer
-;;   :type 'boolean)
-
-(defcustom ess-developer-enter-hook nil
-  "Normal hook run on entering `ess-developer' mode."
-  :group 'ess-developer
-  :type 'hook)
+project root.")
 
-(defcustom ess-developer-exit-hook nil
-  "Normal hook run on exiting `ess-developer' mode."
-  :group 'ess-developer
-  :type 'hook)
+(defvar ess-developer-check-all-dirs t
```

Comment:
Yes. But the current path will contain one of those dirs in the list for sure, and you can start looking from that directory down, no need to descend from current file.